include/wil/registry.h

```diff
@@ -17,16 +17,2855 @@
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param subkey A string to append to the HKEY to attempt to read from
+         *        can be nullptr if not needed
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \return The value read from the registry value of the template type std::optional<DWORD>
+         *         note - will return std::nullopt if the value does not exist
+         * \exception std::exception (including wil::ResultException) will be thrown on failures except value not found
+         */
+        inline ::std::optional<DWORD> try_get_value_dword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name)
+        {
+            return ::wil::reg::try_get_value<DWORD>(key, subkey, value_name);
+        }
+
+        /**
+         * \brief Attempts to read a REGD_DWORD value under a specified key, returning the value in a std::optional
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \return The value read from the registry value of the template type std::optional<DWORD>
+         *         note - will return std::nullopt if the value does not exist
+         * \exception std::exception (including wil::ResultException) will be thrown on failures except value not found
+         */
+        inline ::std::optional<DWORD> try_get_value_dword(HKEY key, _In_opt_ PCWSTR value_name)
+        {
+            return ::wil::reg::try_get_value<DWORD>(key, nullptr, value_name);
+        }
+
+        /**
+         * \brief Attempts to read a REG_QWORD value under a specified key, returning the value in a std::optional
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param subkey A string to append to the HKEY to attempt to read from
+         *        can be nullptr if not needed
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \return The value read from the registry value of the template type std::optional<DWORD64>
+         *         note - will return std::nullopt if the value does not exist
+         * \exception std::exception (including wil::ResultException) will be thrown on failures except value not found
+         */
+        inline ::std::optional<DWORD64> try_get_value_qword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name)

```

Comment:
I debated on that for a while. But now that I'm looking through this entire file, I see I'm inconsistent and use the C++ types elsewhere. I'll just be consistent and use C++ types. Thanks - fixing.

Reply:
heh. turns out DWORD* is not uint32_t*, so I have to fix stuff up a bit to allow both.

Reply:
We made several changes to support both types. We now use STD types like `uint32_t` as the default return type in general, since they can seamlessy decay to the SHOUTY versions (`uint32_t` can be assigned to a `DWORD`, since `DWORD` is a looser type).